content/client-lib-development-guide/features.textile

```diff
@@ -1059,6 +1063,21 @@ h3(#push-channels). Push channels
 *** @(RSH7d2)@ Performs a DELETE request to "/push/channelSubscriptions":/rest-api#delete-channel-subscription with the device's @clientId@ and the channel name.
 ** @(RSH7e)@ @#listSubscriptions(params)@ performs a GET request to "/push/channelSubscriptions":/rest-api#list-channel-subscriptions and returns a paginated result with @PushChannelSubscription@ objects filtered by the provided params, the channel name, the device ID, and the client ID if it exists, as supported by the REST API. A @concatFilters@ param needs to be set to @true@ as well.
 
+h3(#local-device). LocalDevice
+
+* @(RSH8)@ In platforms that support receiving push notifications, the @device()@ method on the @AblyRest@ interface returns an instance of @LocalDevice@ that represents the current state of the device in respect of it being a target for push notifications.
+** @(RSH8a)@ The @LocalDevice@ is initialised when first required, either as a result of a call to @AblyRest.device()@, or as a result of an operation involving the Activation State Machine. The @LocalDevice@ @id@, @clientId@, @deviceSecret@ and @deviceIdentityToken@ attributes are populated, together with any @recipient@-related attributes, to the extent that they exist, from the persisted state.
+** @(RSH8b)@ The @LocalDevice@ @id@ and @deviceSecret@ attributes are generated, and persisted as part of the @LocalDevice@ state, when required by step @(RSH3a2b)@ in the Activation State Machine. At that time, the @clientId@ attribute is also initialised, if the client is identified according to @(RSA7)@.
+** @(RSH8c)@ Following successful registration of a @LocalDevice@, following the procedure in @(RSH3c2a)@, the now known @deviceIdentityToken@ is set and persisted.
+** @(RSH8d)@ If the @LocalDevice* is created by an unidentified client (see @(RSA7)@) and therefore has no @clientId@ set, but the client subsequently becomes identified (as a result of @(RSA7b2)@ or @(RSA7b3)@), then the @LocalDevice@ @clientId@ is set and persisted.
+** @(RSH8e)@ If the @LocalDevice@ @clientId@ becomes set as a result of @(RSH8d)@, and the @LocalDevice@ is already registered (ie the @deviceIdentityToken@ is set), and the ActivationStateMachine is in any state other than @NotActivated@, then , a @GotPushDeviceDetails@ event is sent to "the state machine":#RSH3 containing updated @clientId@.
+** @(RSH8f)@ If the @LocalDevice* is created by an unidentified client (see @(RSA7)@) and therefore has no @clientId@ set, but on receipt of a registration response (see @(RSH3c2)@) the registered device has a non-empty @clientId@, then the @LocalDevice@ @clientId@ is set with that @clientId@.
+** @(RSH8g)@ When validating an existing registration (according to @(RSH3a2a)@), the @LocalDevice@ has a non-empty @clientId@, and the present identified client has a different @clientId@, then activation of the state machine must fail with an error with @code@ 61002.
```

Comment:
> the present identified client has a different @clientId@

Including a null one? Maybe make that explicit?

Also, why have both RSH3a2a1 and RSH8g, each one giving part of the full picture, instead of a single one?

Reply:
> Including a null one? Maybe make that explicit?
Yes, it says "identified" to mean non-null, but I've made that explicit.

> Also, why have both RSH3a2a1 and RSH8g, each one giving part of the full picture, instead of a single one?
Well I wanted to ensure all requirements related to @LocalDevice@ were now in one place, but I agree it could be clearer.

Reply:
https://github.com/ably/docs/pull/786/commits/f7115c4b41dceb108b2f5eb25468fb96bdcdaaa7